File: shared/types/landmarks.ts
/**
 * Shared TypeScript definitions for MediaPipe Holistic landmarks.
 *
 * These types define the contract between clients (Flutter/React) and the
 * recognition service for holistic landmark data (pose + hands).
 *
 * MediaPipe returns normalized coordinates in the range [0, 1] for x, y, z.
 * Visibility represents confidence of landmark detection (0-1).
 *
 * Total dimensions: 67 landmarks × 3D = 201 features
 * - Pose: 25 landmarks × 3 = 75 features
 * - Left hand: 21 landmarks × 3 = 63 features
 * - Right hand: 21 landmarks × 3 = 63 features
 */

/**
 * Single 3D landmark point from MediaPipe.
 * Coordinates are normalized [0, 1] relative to image dimensions.
 */
export interface Landmark {
  x: number;
  y: number;
  z: number;
  visibility: number;
}

/**
 * Hand landmarks container.
 * MediaPipe detects 21 landmarks per hand in standard format.
 */
export interface HandLandmarks {
  /** 'Left' or 'Right' - handedness of the detected hand */
  handedness: 'Left' | 'Right';
  /** Array of exactly 21 landmark points */
  landmarks: Landmark[];
}

/**
 * Pose landmarks container.
 * MediaPipe Holistic detects 25 upper body pose landmarks.
 * Indices: 0=nose, 1=left eye inner, 2=left eye, 3=left eye outer,
 * 4=right eye inner, 5=right eye, 6=right eye outer, 7=left ear,
 * 8=right ear, 9=mouth left, 10=mouth right, 11=left shoulder,
 * 12=right shoulder, 13=left elbow, 14=right elbow, 15=left wrist,
 * 16=right wrist, 17=left pinky, 18=right pinky, 19=left index,
 * 20=right index, 21=left thumb, 22=right thumb, 23=left hip,
 * 24=right hip
 */
export interface PoseLandmarks {
  /** Array of exactly 25 landmark points (upper body) */
  landmarks: Landmark[];
}

/**
 * Complete holistic landmarks payload from client to recognition service.
 * Contains pose and hand landmarks (any may be optional for graceful degradation).
 *
 * Feature vector format (201 dimensions):
 * [pose25×3, left_hand21×3, right_hand21×3]
 */
export interface HolisticLandmarksPayload {
  /** Pose landmarks (upper body - optional for backward compatibility) */
  pose?: PoseLandmarks;

  /** Left hand landmarks (optional - may not be visible) */
  left?: HandLandmarks;

  /** Right hand landmarks (optional - may not be visible) */
  right?: HandLandmarks;

  /** Unix timestamp in milliseconds when frame was captured */
  timestamp: number;

  /** Unique session identifier for the recognition session */
  sessionId: string;
}

/**
 * @deprecated Use HolisticLandmarksPayload for new implementations.
 * Kept for backward compatibility with existing hand-only pipelines.
 */
export type HandLandmarksPayload = HolisticLandmarksPayload;

File: recognition-service/app/schemas/landmarks.py
"""
Pydantic schemas for landmark payloads.
Supporting MediaPipe Holistic output (pose + hands).
"""

from pydantic import BaseModel, field_validator
from typing import Optional, Dict, Any, List


class Landmark(BaseModel):
    """Single 3D landmark point."""
    x: float
    y: float
    z: float
    visibility: float = 1.0


class HandLandmarks(BaseModel):
    """Hand landmarks with handedness."""
    handedness: str
    landmarks: List[Landmark]

    @field_validator('landmarks')
    @classmethod
    def validate_hand_landmarks_count(cls, v):
        """Validate exactly 21 hand landmarks."""
        if len(v) != 21:
            raise ValueError(f'Hand must have exactly 21 landmarks, got {len(v)}')
        return v


class PoseLandmarks(BaseModel):
    """MediaPipe pose landmarks."""
    landmarks: List[Landmark]

    @field_validator('landmarks')
    @classmethod
    def validate_pose_landmarks_count(cls, v):
        """Validate exactly 33 MediaPipe pose landmarks."""
        if len(v) != 33:
            raise ValueError(f'Pose must have exactly 33 landmarks, got {len(v)}')
        return v


class LandmarksPayload(BaseModel):
    """Complete landmarks payload from client."""
    landmarks: Dict[str, Any]
    timestamp: int
    sessionId: str

    @field_validator('landmarks')
    @classmethod
    def validate_landmarks_structure(cls, v):
        """Validate landmarks dict has proper structure."""
        if not isinstance(v, dict):
            raise ValueError('landmarks must be a dictionary')

        # At least one of pose, left, or right must be present
        has_pose = 'pose' in v and v['pose'] is not None
        has_left = 'left' in v and v['left'] is not None
        has_right = 'right' in v and v['right'] is not None

        if not (has_pose or has_left or has_right):
            raise ValueError('At least one of pose, left, or right landmarks required')

        # Validate pose if present
        if has_pose:
            pose_data = v['pose']
            if not isinstance(pose_data, dict):
                raise ValueError('pose must be a dictionary')
            if 'landmarks' not in pose_data:
                raise ValueError('pose missing landmarks')
            if len(pose_data['landmarks']) != 33:
                raise ValueError(f'pose must have exactly 33 landmarks, got {len(pose_data["landmarks"])}')

        # Validate left hand if present
        if has_left:
            hand = v['left']
            if not isinstance(hand, dict):
                raise ValueError('left must be a dictionary')
            if 'handedness' not in hand:
                raise ValueError('left missing handedness')
            if 'landmarks' not in hand:
                raise ValueError('left missing landmarks')
            if len(hand['landmarks']) != 21:
                raise ValueError(f'left must have exactly 21 landmarks, got {len(hand["landmarks"])}')

        # Validate right hand if present
        if has_right:
            hand = v['right']
            if not isinstance(hand, dict):
                raise ValueError('right must be a dictionary')
            if 'handedness' not in hand:
                raise ValueError('right missing handedness')
            if 'landmarks' not in hand:
                raise ValueError('right missing landmarks')
            if len(hand['landmarks']) != 21:
                raise ValueError(f'right must have exactly 21 landmarks, got {len(hand["landmarks"])}')

        return v
